### ISSUE TITLE
Add special properties for relative values (`ThreadsProperty`) 

### ISSUE THREAD

**Issue by leonard84** [gradle/gradle#21394]
## Summary
Having the ability to set relative values for Memory/CPU would be really helpful as fixed values are not really portable, being either too much for small machines or leaving resources unused on larger machines. So users end up writing the same helper methods again and again.

### Expected Behavior

A special subclass of `Property<Integer>`, e.g., `ThreadsProperty`, is added that offers methods like.
* `quarterCpus()`
* `halfCpus()`
* `threeQuartersCpus()`
* `allCpus()`
* `partialCpus(fraction:Double)` calculates the used threads based on the given fraction, with the minimum of 1
* `partialCpusWithReserve(fraction:Double, reserve:Int)` similar to `partialCpus` but will make sure that at least the number `reserve` threads are not used

A similar property could be added for memory, e.g., a `MemoryProperty`.

### Current Behavior

Currently, there is no such support built-in, and users have to manually calculate this based on `Runtime.getRuntime().availableProcessors()`, or use a fixed value.

### Context

A problem is that tasks can be executed in parallel, and these properties don't talk to each other.
I'd consider it out of scope and optional future work to solve the problem of managing concurrent resource usage.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

While the intent makes sense, we are not sure if it should be implemented as a special-type property.

**Comment by bamboo** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.

***

The rich modeling makes sense. However, it does not require new property types. It calls for new value types (e.g. `CpuAllocation` and `Memory`), factory functions and constants (e.g. `CpuAllocation.HALF_CPUS`) which can be composed with the existing property type (e.g. `Property<CpuAllocation>` with say `convention(CpuAllocation.HALF_CPUS)`).
